src/java/org/apache/cassandra/index/sai/StorageAttachedIndexGroup.java

```diff
@@ -250,31 +251,61 @@ public boolean handles(IndexTransaction.Type type)
     }
 
     @Override
-    public Set<Component> getComponents()
+    public Set<Component> componentsForNewBuid(Descriptor descriptor, TableMetadata metadata)
     {
-        return getComponents(indices);
+        return componentsForNewBuid(descriptor, metadata, indices);
     }
 
-    static Set<Component> getComponents(Collection<StorageAttachedIndex> indices)
+    static Set<Component> componentsForNewBuid(Descriptor descriptor, TableMetadata metadata, Collection<StorageAttachedIndex> indices)
     {
-        Set<Component> components = Version.latest().onDiskFormat()
-                                                  .perSSTableComponents()
-                                                  .stream()
-                                                  .map(c -> new Component(Component.Type.CUSTOM,
-                                                                          Version.latest().fileNameFormatter().format(c, null)))
-                                                  .collect(Collectors.toSet());
-        indices.forEach(index -> components.addAll(index.getComponents()));
+        // The components created depends, amongst other things, on the generation that needs to be used for the new
+        // build (which depends on the version, but that's always the latest for new builds). That generation is always
+        // 0 for new sstables, but may not be for rebuilds. To handle that, we rely on the same logic used when we do
+        // write the new components (`IndexDescriptor.newPerSSTableGroupWriter`/`IndexDescriptor.newPerIndexGroupWriter`),
+        // even if we don't write anything at that point.
+        // Do note that we create an `IndexDescriptor` from scratch, and do not look in the context manager for an
```

Comment:
Oops, classic me: I started the comment, though of something else, and forgot to finish it...

Pushed the end of the comment, but yes, the idea here is that this is call any time we write index components, and so half the time it will be for sstables that don't exist yet (flush or compaction). And I don't want to populate the descriptor in the `SSTableContextManager` just yet because if something fails, it wouldn't be cleared correctly.

Admittedly, half of the time the sstable will exists (rebuild, or just building a new index on an existing sstable) so we could try go get the descriptor from the context only if it exists and call `IndexDescriptor.create` directly otherwise, but this didn't felt worth the added code.